apollo-router/src/registry/mod.rs

```diff
@@ -0,0 +1,175 @@
+use thiserror::Error;
+
+/// Configuration for fetching an OCI Bundle
+/// This struct does not change on router reloads - they are all sourced from CLI options.
+#[derive(Debug, Clone)]
+pub struct OCIConfig {
+    /// The Apollo key: `<YOUR_GRAPH_API_KEY>`
+    pub apollo_key: String,
+
+    // Should this be a reference instead?
+    pub url: String,
+}
+
+pub(crate) struct OCIResult {
+    pub schema: String,
+}
+
+#[derive(Debug, Error)]
+pub(crate) enum Error {
+    #[error("OCI layer has not title")]
+    OCILayerMissingTitle,
+    #[error("Oci Distribution error: {0}")]
+    OCIDistributionError(OciDistributionError),
+    #[error("Oci Parsing error: {0}")]
+    OCIParseError(oci_client::ParseError),
+    #[error("Unable to parse layer: {0}")]
+    OCILayerParseError(FromUtf8Error),
+}
+
+impl From<oci_client::ParseError> for Error {
+    fn from(value: oci_client::ParseError) -> Self {
+        Error::OCIParseError(value)
+    }
+}
+
+impl From<OciDistributionError> for Error {
+    fn from(value: OciDistributionError) -> Self {
+        Error::OCIDistributionError(value)
+    }
+}
+
+impl From<FromUtf8Error> for Error {
+    fn from(value: FromUtf8Error) -> Self {
+        Error::OCILayerParseError(value)
+    }
+}
+
+fn build_auth(reference: &Reference, apollo_key: &String) -> RegistryAuth {
+    let server = reference
+        .resolve_registry()
+        .strip_suffix('/')
+        .unwrap_or_else(|| reference.resolve_registry());
+
+    // Check if the server registry ends with apollographql.com
+    if server.ends_with("apollographql.com") {
+        tracing::debug!("Using Apollo registry authentication");
+        return RegistryAuth::Basic("apollo_registry".to_string(), apollo_key.clone());
+    }
+
+    match docker_credential::get_credential(server) {
+        Err(CredentialRetrievalError::ConfigNotFound) => RegistryAuth::Anonymous,
+        Err(CredentialRetrievalError::NoCredentialConfigured) => RegistryAuth::Anonymous,
+        Err(e) => {
+            tracing::warn!("Error handling docker configuration file: {}", e);
+            RegistryAuth::Anonymous
+        }
+        Ok(DockerCredential::UsernamePassword(username, password)) => {
+            tracing::debug!("Found docker credentials");
+            RegistryAuth::Basic(username, password)
+        }
+        Ok(DockerCredential::IdentityToken(_)) => {
+            tracing::warn!(
+                "Cannot use contents of docker config, identity token not supported. Using anonymous auth"
+            );
+            RegistryAuth::Anonymous
+        }
+    }
+}
+
+async fn pull_oci(
+    client: &mut ociClient,
+    auth: &RegistryAuth,
+    reference: &Reference,
+) -> Result<OCIResult, Error> {
+    tracing::info!(?reference, "pulling oci bundle");
+    let supported_types = vec![
+        "application/json",
+        "text/plain",
+        "application/vnd.oci.empty.v1+json",
+    ];
+    let image = client
+        .pull(reference, auth, supported_types.clone())
```

Comment:
You don't need the clone because you aren't writing layers.

```suggestion
        .pull(reference, auth, supported_types)
```